sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -308,22 +317,29 @@ protected static String buildTemporaryFilename(String tempDirectory, String file
      * @param sink the FileBasedSink that will be used to configure this write operation.
      */
     public FileBasedWriteOperation(FileBasedSink<T> sink) {
-      this(sink, buildTemporaryDirectoryName(sink.getBaseOutputFilename()));
+      this(sink, NestedValueProvider.of(
+          sink.getBaseOutputFilenameProvider(), new TemporaryDirectoryBuilder()));
     }
 
-    private static String buildTemporaryDirectoryName(String baseOutputFilename) {
-      try {
-        IOChannelFactory factory = IOChannelUtils.getFactory(baseOutputFilename);
-        Path baseOutputPath = factory.toPath(baseOutputFilename);
-        return baseOutputPath
-            .resolveSibling(
-                "temp-beam-"
-                    + baseOutputPath.getFileName()
-                    + "-"
-                    + Instant.now().toString(DateTimeFormat.forPattern("yyyy-MM-DD_HH-mm-ss")))
-            .toString();
-      } catch (IOException e) {
-        throw new RuntimeException(e);
+    private static class TemporaryDirectoryBuilder
+        implements SerializableFunction<String, String> {
+      Instant now = Instant.now();
```

Comment:
The intent of the code is to have a consistent value of tempDirectory across all workers, which wouldn't happen if now() was called inline.  The NVP is necessary because we can't access baseOutputFilename at construction time.

I think it's fine that this is the same for all invocations of the template, since jobs shouldn't be sharing an output directory.